### DISCUSSION TITLE
TypeError: expected str, bytes or os.PathLike object, not BytesIO

### DISCUSSION THREAD

**Discussion by hellwraiz** [pyinstaller/pyinstaller#6133] [Help]
Everything was fine before I used `pyinstaller` to convert my file to a .exe format, but after that, I received this error:

`Traceback (most recent call last):`
`  File "v1.1.py", line 610, in <module>`
`  File "v1.1.py", line 57, in the_main_loop`
`  File "v1.1.py", line 268, in getting_boxes`
`TypeError: expected str, bytes or os.PathLike object, not BytesIO`
`[15696] Failed to execute script 'v1.1' due to unhandled exception!`

And following its directions, it lead me to this line:

`font = pygame.font.Font(None, position_size['extras']['font_size'])`

where position_size is a dictionary where I store all sizes and positions of all things and here is the bit I used:

`position_size{
    'extras':{
        'font_size':36
}
}`

Which once again was absolutely fine before I used `pyinstaller`. Is there anything I can do to fix this? I really need to convert this app to .exe.

**Comment by bwoodsend** [comment]
Can you give us an example piece of code to test on. All we know from what you've given us is that you're using pygame but a basic pygame example works fine.

**Reply by rokm** [reply]
Here's the reproducer:

```python
import pygame
pygame.init()
font = pygame.font.Font(None, 36)
print(font)
```

Looks like that `None` is somehow handled differently in the frozen application...

**Reply by rokm** [reply]
The upstream issue is pygame/pygame#2603, and a work-around/fix was already merged (pygame/pygame#2604). However, it has not been released yet and requires installing from their git `main`.

**Reply by rokm** [reply]
And a snippet that simulates what is going on in `pygame.font` extension:
```python
import pygame
pygame.init()

import pygame.pkgdata
font_data = pygame.pkgdata.getResource("freesansbold.ttf")  # _io.BufferedReader in unfrozen script, _io.BytesIO in frozen application
```

And the "unexpected" `io.BytesIO` comes from our [`pkg_resources` provider](https://github.com/pyinstaller/pyinstaller/blob/f7c8291d17b07ed00ccc27b836e04319947bdacc/PyInstaller/hooks/rthooks/pyi_rth_pkgres.py#L95) not implementing `get_resource_stream` method. Therefore, the parent `pkg_resources.NullProvider` is used, which in turn ends up calling our [FrozenImporter's get_data method](https://github.com/pyinstaller/pyinstaller/blob/f7c8291d17b07ed00ccc27b836e04319947bdacc/PyInstaller/loader/pyimod03_importers.py#L305-L324). 

We *could* work around this problem by implementing `get_resource_stream` that returns a file handle (`open(name, 'rb')`) for on-disk resources and fall back to parent implementation for embedded ones (or even directly call loader's `get_data`). However, the current behavior is conformant to spec, which requires us to return a file-like object, not a file handle. So I'm inclined to leave this as is, and have `pygame` fix their behavior on their side (which they will do in their next release).
